### ISSUE TITLE
label-has-associated-control error with label/input

### ISSUE THREAD

**Issue by fritogotlayed** [evcohen/eslint-plugin-jsx-a11y#632]
I believe I'm seeing an issue with the following snippet on `eslint-plugin-jsx-a11y 6.2.3`. I believe we're going to disable the rule for now and just re-enable after it gets resolved.

```
          <div className="field">
            <label htmlFor="nameInput" className="label">Name</label>
            <div className="control has-icons-left">
              <input
                id="nameInput"
                className="input"
                disabled={isPosting}
                type="text"
                name="name"
                placeholder="How should we address you"
                value={name}
                onChange={this.onChange}
              />
              <span className="icon is-small is-left">
                <i className="fas fa-tag" />
              </span>
            </div>
          </div>
```

[Output from the run config](https://gist.github.com/fritogotlayed/ac7d1ab14b09c987da85f4ae0684288a) appears that this case should be covered and function properly via the [documentation](https://github.com/evcohen/eslint-plugin-jsx-a11y/blob/master/docs/rules/label-has-associated-control.md#case-the-label-is-a-sibling-of-the-control) and [conversation in another issue I initially read as identical](https://github.com/evcohen/eslint-plugin-jsx-a11y/issues/566). From what I'm seeing in the output the depth here is 25 so I'm confused as why this is not working as I believe it should. 

ESLint run output snippet:
```
/[path]/src/screens/RegisterScreen.jsx
  153:13  error  A form label must be associated with a control  jsx-a11y/label-has-associated-control
  172:13  error  A form label must be associated with a control  jsx-a11y/label-has-associated-control
  193:13  error  A form label must be associated with a control  jsx-a11y/label-has-associated-control
  213:13  error  A form label must be associated with a control  jsx-a11y/label-has-associated-control

```

Relevant snippet from eslint run rules: 
```
    "jsx-a11y/label-has-associated-control": [
      "error",
      {
        "labelComponents": [],
        "labelAttributes": [],
        "controlComponents": [],
        "assert": "both",
        "depth": 25
      }
    ],
```

**Comment by ljharb** [comment]
The "both" assert means that the input has to also be inside the label. If you change it to "either" or "htmlFor", it will permit your example.

**Comment by fritogotlayed** [comment]
@ljharb You sir are a rockstar! I apologize for creating a bit of distraction on the other issue. Switching from "both" to "either" knocked it right out. For anyone who may stumble here and is also using the yml format: 

```
rules:
  jsx-a11y/label-has-associated-control:
    - error
    - labelComponents: []
      labelAttributes: []
      controlComponents: []
      assert: either
      depth: 25
```

**Comment by ljharb** [comment]
I would, however, suggest nesting the input inside the label, since doing both covers more assistive devices and allows the maximal amount of people to use the input properly.

**Comment by hades200082** [comment]
One of the most common UI frameworks out there, Bootstrap, uses the `for=""` method and does not use nesting in any of their documented examples.

https://getbootstrap.com/docs/4.5/components/forms/

From their documentation... https://getbootstrap.com/docs/4.5/getting-started/accessibility/

> The overall accessibility of any project built with Bootstrap depends in large part on the author’s markup, additional styling, and scripting they’ve included. However, provided that these have been implemented correctly, it should be perfectly possible to create websites and applications with Bootstrap that fulfill WCAG 2.0 (A/AA/AAA), Section 508 and similar accessibility standards and requirements.

This seems to indicate that both are indeed NOT required when using the `for=""` form and that you can still achieve a highly accessible website without nesting.

**Comment by ljharb** [comment]
@hades200082 on the contrary; that just indicates that bootstrap isn't concerned with the devices and human beings (hearing-impaired or otherwise) that excludes - including the latest version of Dragon Naturally Speaking, a commonly used tool.

(also, please don't cross-post the same comment both here and [here](https://github.com/airbnb/javascript/pull/2136), it's noise)

**Comment by hades200082** [comment]
If Bootstrap can support WCAG 2.0 AAA as indicated by their documentation how are they not concerned?

**Comment by ljharb** [comment]
Because the standard isn’t actually relevant - what matters is what actual devices out there in usage are following. There are non standards compliant devices in wide use, thus, to be accessible one must match the devices, not the standard.

**Comment by hades200082** [comment]
It sounds more like you have a personal axe to grind on this subject. By continuing to support outdated and non-compliant devices you're perpetuating their continued production and sale/usage.

The biggest obstacle in many businesses (particularly SMEs) to building accessible websites is the cost of development. Indeed, issues such as this one which take more skill and development effort to overcome add to that cost.

Surely it would be better to guide the industry that is making such devices towards adopting standards that make developing accessible websites easier so that more companies & developers can do so within their budgets?

By your current argument all websites should still be using only HTML/CSS/JS that is supported by IE6 because there are a large number of people in the world that are still stuck with that browser.

**Comment by ljharb** [comment]
If my personal axe is “not excluding human beings”, then you’re right :-)

Supporting old devices/platforms is critical for those that can’t update. Not supporting them is hostile. Nobody’s using an old platform because they want to. As for non-compliance, again, hurting people isn’t a just way to force a corporation to change.

Adding both nesting and for-ID linking isn’t a large cost. Airbnb has 30,000+ react components (as of last July) and all form controls use this approach.

There is nobody still using IE 6 because it can’t even connect to SSL anymore. You should be supporting IE 9, ideally, but that depends on your demographic’s usage. The difference here is that a hearing-impaired user on modern Chrome will still be screwed if you don’t follow this recommendation - why are you advocating to discriminate against them? (note that US law also allows massive financial penalties in civil court for discriminating against protected classes, and this exact category already applies, per court precedent)

**Comment by hades200082** [comment]
> The difference here is that a hearing-impaired user on modern Chrome will still be screwed if you don’t follow this recommendation

I didn't realise that hearing had anything to do with ability to read or use a website form?

> why are you advocating to discriminate against them?

I'm not advocating or discriminating against anyone. I'm stating that standards exist to allow companies and developers to provide for these types of people in a way that is economically and technically viable for smaller businesses that can't spend millions on developers.

By enforcing your very narrow view all that will happen is that those companies and their developers will, at best, simply ignore the warnings your package throws up and at worst, remove the package or disable it.

Rather than fighting a crusade that will lead to the opposite of your intended goal, why not work with the established standards and use your knowledge and influence to work with the people defining the standards to improve them. Likewise, working with the developers of assistive technology to help them to better adhere to those standards.

The web became a better place, both for users and for developers when the W3C was formed and browser vendors started working together to produce comparable user experiences across different devices and browsers. Why should assistive tech be any different?

**Comment by hades200082** [comment]
It's also worth noting that I have checked the websites of several accessible browsers and none of them follow this recommendation themselves.

Check out ...

- the donation form on: https://www.nvaccess.org/download/
- Apple's purchasing form (they advocate the Mac assistive tech VoiceOver on their devices): https://www.apple.com/uk/shop/buy-mac/imac/21.5-inch-3.6ghz-quad-core-processor-1tb
- The contact form on the website of Dragon Naturally Speaking (which you mentioned above) doesn't even use labels, only placeholders - https://www.nuance.com/en-gb/dragon/engage-us.html
- Project Orca - the search box (the only form on the page) doesn't use nesting - https://wiki.gnome.org/Projects/Orca
- Search box on Airbnb help centre & resource centre uses no label, only placeholder - https://www.airbnb.co.uk/help/home & https://www.airbnb.co.uk/resources/hosting-homes
- Location dropdown box on Airbnb careers site has a h3 as it's label, not nested. - https://careers.airbnb.com/
- job application forms on Airbnb careers site do not use nesting for labels - https://careers.airbnb.com/contractors/1131934/?application=true

**Comment by ljharb** [comment]
Sorry, you're right; i meant vision-impaired. My mistake.

That many websites do a poor job of accessibility highlights why it's so important to maintain high standards - it's not an excuse for lax ones.

**Comment by qkreltms** [comment]
Seems not working when you use ```styled-component```

**Comment by panveel** [comment]
Seems not working when you use styled-component any ideas?

**Comment by panveel** [comment]
> Seems not working when you use `styled-component` any ideas?

Having the same issue with `styled-component`.

**Comment by saikrishnabanda5** [comment]
COuld you tell me what to do for it?
